ios/LayoutReanimation/REASharedTransitionManager.m

```diff
@@ -0,0 +1,512 @@
+                  with:@selector(swizzled_viewDidLayoutSubviews)
+              forClass:[viewController class]];
+    [self swizzlMethod:@selector(notifyWillDisappear)
+                  with:@selector(swizzled_notifyWillDisappear)
+              forClass:[view class]];
+  });
+}
+
+- (void)swizzlMethod:(SEL)originalSelector with:(SEL)swizzledSelector forClass:(Class)originalClass
+{
+  Class class = [self class];
+  Method originalMethod = class_getInstanceMethod(originalClass, originalSelector);
+  Method swizzledMethod = class_getInstanceMethod(class, swizzledSelector);
+  IMP originalImp = method_getImplementation(originalMethod);
+  IMP swizzledImp = method_getImplementation(swizzledMethod);
+  class_replaceMethod(originalClass, swizzledSelector, originalImp, method_getTypeEncoding(originalMethod));
+  class_replaceMethod(originalClass, originalSelector, swizzledImp, method_getTypeEncoding(swizzledSelector));
+}
+
+- (void)swizzled_viewDidLayoutSubviews
+{
+  /*
+    The screen header is added later than other screen children and changes their
+    position on the screen. That's why we need to make a snapshot after the header
+    mount. We don't want to add a strong dependency to the react-native-screens
+    library so we decided to use KVO from Obj-C. However, all properties of the
+    screen are updated without calling Obj-C's setters. We swizzled the method
+    `viewDidLayoutSubviews` - this method is called after the appearance of the header.
+    Then we call KVO to notify Reanimated observer about the attached header.
+  */
+
+  // call original method from react-native-screens, self == RNScreen
+  [self swizzled_viewDidLayoutSubviews];
+  // get RNScreenView from RNScreen to call KVO on RNScreenView because
+  // RNScreen doesn't contains writable public property.
+  UIView *view = [self valueForKey:@"screenView"];
+  // call KVO to run runAsyncSharedTransition from Reanimated
+  [view setValue:[view valueForKey:@"transitionDuration"] forKey:@"transitionDuration"];
+}
+
+- (void)swizzled_notifyWillDisappear
+{
+  // call original method from react-native-screens, self == RNScreenView
+  [self swizzled_notifyWillDisappear];
+  // call KVO to run runAsyncSharedTransition from Reanimated
+  [self setValue:[self valueForKey:@"activityState"] forKey:@"activityState"];
+}
+
+- (void)observeValueForKeyPath:(NSString *)keyPath
+                      ofObject:(id)object
+                        change:(NSDictionary<NSKeyValueChangeKey, id> *)change
+                       context:(void *)context
+{
+  UIView *screen = (UIView *)object;
+  UIView *stack = screen;
+  while (stack != nil && ![stack isKindOfClass:[RNSScreenStackView class]]) {
+    stack = stack.superview;
+  }
+
+  if ([keyPath isEqualToString:@"transitionDuration"]) {
+    // added new screen (push)
+    if (screen.superview != nil) {
+      [self runAsyncSharedTransition];
+    }
+  } else if ([keyPath isEqualToString:@"activityState"]) {
```

Comment:
`activityState` is ever changed in `native-stack`? I thought it is only used in other navigators under `RNSScreenContainerView`.

Reply:
You're right but I'm changing it in `swizzled_notifyWillDisappear` method.